text/0731-set-route-component.md

```diff
@@ -0,0 +1,170 @@
+format, but also partially because you cannot use them currently with a very
+important part of Ember: Routes.
+
+There currently is not a way to directly set the template of a route the same
+way that you can set the template of a component with `setComponentTemplate`.
+This means you cannot associate a template defined with `precompileTemplate`
+with a route, and so there is no way to use imported values with them. Instead,
+components must still be defined in the `app/components` directory, and then
+used via resolution in a route's template defined in `app/templates/routes`.
+
+This RFC seeks to add a method for associating a _component_ with a route. This
+will allow users to associate strict mode templates with a route via a
+component - either a template only component or one with a backing class. This
+will allow the community to continue experimenting with different template
+import formats more thoroughly, and will also allow users to experiment with
+building Ember apps without controllers by effectively bypassing them.
+Controllers will not be fully removed, as they still have some functionality
+which has not been replaced, but this will help to see where those gaps are and
+figure out what is needed to fill them.
+
+This takes us one step further into the pit of incoherence, with the goal of
+better defining how to move forward in two major areas that we are aiming for in
+the next peak of coherence.
+
+## Detailed design
+
+This RFC proposes adding the `setRouteComponent` API, importable from
+`@ember/routing/route`.
+
+```js
+declare function setRouteComponent(component: Component, route: Route): void;
+```
+
+Users can associate any component definition with a route via
+`setRouteComponent`. When a component has been associated with a route this way,
+the route will render this component at its root rather than rendering the route
+template defined in `app/templates/routes`. The component will be passed the
+`@model` and `@controller` arguments, which correspond to the model returned
+from the route's `model` hook and the instance of the route's controller.
+
+This will effectively be syntactic sugar for rendering a top-level component
+within a route template today:
+
+```js
+import Route, { setRouteComponent } from '@ember/routing/route';
+import MyRouteComponent from './component';
+
+export default class MyRoute extends Route {}
+
+setRouteComponent(MyRouteComponent, MyRoute);
+```
+
+Is equivalent to having the following in the `my-route.hbs` file:
+
+```hbs
+<MyRouteComponent @model={{@model}} @controller={{this}} />
+```
+
+Assuming that `MyRouteComponent` is resolvable via that name.
+
+No changes will be made to the route or controller lifecycle, hooks, or
+behaviors in general, since this is not related to the goal of unlocking
+experimentation with strict mode and route templates, or the goal of
+experimenting with controller-less applications. Changes in these areas will be
+done in future RFCs instead.
```

Comment:
One unanswered question I thought of here: how does this interact with the existing abilities to muck with specific rendering patterns on routes, [`render`](https://api.emberjs.com/ember/3.25/classes/Route/methods/render?anchor=render) and [`renderTemplate`](https://api.emberjs.com/ember/3.25/classes/Route/methods/renderTemplate?anchor=renderTemplate)? Is using those with this a hard error, a deprecation, a warning, etc.? And what's the actual behavior—is `renderTemplate` ignored if using `setRouteComponent`?

Reply:
Ah yes, we should address that. I believe these APIs would be ignored, since we are fully ignoring the route's template in general. Since these APIs are [deprecated](https://deprecations.emberjs.com/v3.x#toc_route-render-template), I think this also won't introduce any inconsistencies.